storey/redis_driver.py

```diff
@@ -0,0 +1,482 @@
+
+import redis
+import rediscluster
+from .dtypes import RedisError
+from .utils import schema_file_name
+
+
+class RedisType(Enum):
+    STANDALONE = 1
+    CLUSTER = 2
+
+
+class NeedsRedisAccess:
+    """Checks that params for access to Redis exist and are legal
+
+    :param webapi: URL to the web API (https or http). If not set, the REDIS_URL environment variable will be used.
+
+    """
+
+    def __init__(self, redis_url=None):
+        redis_url = redis_url or os.getenv('REDIS_URL')
+        if not redis_url:
+            self._redis_url = None
+            print('Missing redis_url parameter or REDIS_URL environment variable. Using fakeredit instead')
+            return
+
+        if not redis_url.startswith('redis://'):
+            redis_url = f'redis://{redis_url}'
+
+        self._redis_url = redis_url
+
+
+class RedisDriver(NeedsRedisAccess, Driver):
+    REDIS_TIMEOUT = 5  # Seconds
+    REDIS_WATCH_INTERVAL = 1  # Seconds
+    DATETIME_FIELD_PREFIX = "_dt:"
+    TIMEDELTA_FIELD_PREFIX = "_td:"
+    DEFAULT_KEY_PREFIX = "storey:"
+    _thread_pool = concurrent.futures.ThreadPoolExecutor()
+
+    def __init__(self,
+                 redis_client: Optional[Union[redis.Redis, rediscluster.RedisCluster]] = None,
+                 redis_type: RedisType = RedisType.STANDALONE,
+                 key_prefix: str = None,
+                 redis_url: Optional[str] = None,
+                 aggregation_attribute_prefix: str = 'aggr_',
+                 aggregation_time_attribute_prefix: str = '_',
+                 use_parallel_operations: bool = True,  # unused
+                 ):
+
+        NeedsRedisAccess.__init__(self, redis_url)
+        self._redis = redis_client
+        self._key_prefix = key_prefix if key_prefix else self.DEFAULT_KEY_PREFIX
+        self._type = redis_type
+        self._mtime_name = '$_mtime_'
+
+        self._aggregation_attribute_prefix = aggregation_attribute_prefix
+        self._aggregation_time_attribute_prefix = aggregation_time_attribute_prefix
+        self._aggregation_prefixes = (self._aggregation_attribute_prefix,
+                                      self._aggregation_time_attribute_prefix)
+
+    def asyncify(self, fn):
+        """Run a synchronous function asynchronously."""
+        async def inner_fn(*args, **kwargs):
+            loop = asyncio.get_event_loop()
+            partial_hset = partial(fn, *args, **kwargs)
+            return await loop.run_in_executor(RedisDriver._thread_pool, partial_hset)
+        return inner_fn
+
+    @property
+    def redis(self):
+        if self._redis:
+            return self._redis
+        if self._type is RedisType.STANDALONE:
+            return redis.Redis.from_url(self._redis_url, decode_responses=True)
+        self._redis = rediscluster.RedisCluster.from_url(self._redis_url, decode_response=True)
+        return self._redis
+
+    @staticmethod
+    def make_key(key_prefix, *parts):
+        return f"{key_prefix}{''.join([str(p) for p in parts])}"
```

Comment:
```suggestion
        key = ''.join([str(p) for p in parts])
        return f'{key_prefix}{key}'
```
but hm, is it even correct to join the keys without a separator?

Reply:
it is the table_path part contains '/' both at it's start and end.